ts-client/src/amm/curve/stable-swap.ts

```diff
@@ -8,85 +8,62 @@ import {
+import JSBI from 'jsbi';
+import { Token, TokenAmount, Percent, ChainId } from '@saberhq/token-utils';
+import { AccountInfo, Connection, Keypair, PublicKey } from '@solana/web3.js';
+import { Idl } from '@project-serum/anchor/dist/esm';
+import MarinadeIDL from '../marinade-finance.json';
+import { CURVE_TYPE_ACCOUNTS } from '../constants';
+import { Depeg, DepegType, PoolFees, TokenMultiplier } from '../types';
+import { getOnchainTime } from '../utils';
 
 // Precision for base pool virtual price
 const PRECISION = new BN(1_000_000);
 const BASE_CACHE_EXPIRE = new BN(60 * 10);
 
 export class StableSwap implements SwapCurve {
-  amp: number;
-  tokenMultiplier: TokenMultiplier;
-  depeg: Depeg;
-  extraAccounts: Map<String, AccountInfo<Buffer>>;
-  onChainTime: number;
-
   constructor(
-    amp: number,
-    tokenMultiplier: TokenMultiplier,
-    depeg: Depeg,
-    onChainTime: number,
-    extraAccounts: Map<String, AccountInfo<Buffer>>
-  ) {
-    this.amp = amp;
-    this.tokenMultiplier = tokenMultiplier;
-    this.depeg = depeg;
-    this.onChainTime = onChainTime;
-    this.extraAccounts = extraAccounts;
-  }
+    private amp: number,
+    private tokenMultiplier: TokenMultiplier,
+    private depeg: Depeg,
+    private extraAccounts: Map<String, AccountInfo<Buffer>>,
+    private connection: Connection,
+  ) {}
 
   private getBasePoolVirtualPrice(depegType: DepegType): BN {
-    if (depegType["marinade"]) {
-      const account = this.extraAccounts.get(
-        EXTRA_ACCOUNTS.marinade[0].toBase58()
-      );
+    if (depegType['marinade']) {
+      const account = this.extraAccounts.get(CURVE_TYPE_ACCOUNTS.marinade.toBase58());
       const coder = new BorshCoder(MarinadeIDL as any as Idl);
-      const stake = coder.accounts.decode("State", account!.data);
+      const stake = coder.accounts.decode('State', account!.data);
       const msolPrice = stake.msolPrice as BN;
       return msolPrice.mul(PRECISION).div(new BN(0x1_0000_0000));
     }
-    if (depegType["lido"]) {
-      const account = this.extraAccounts.get(
-        EXTRA_ACCOUNTS.solido[0].toBase58()
-      );
+    if (depegType['lido']) {
+      const account = this.extraAccounts.get(CURVE_TYPE_ACCOUNTS.solido.toBase58());
       //https://github.com/mercurial-finance/mercurial-dynamic-amm/blob/main/programs/amm/tests/test_depeg_price.rs#L33
       const stSolSupply = new BN(account!.data.readBigInt64LE(73).toString());
       const stSolBalance = new BN(account!.data.readBigInt64LE(81).toString());
       return stSolBalance.mul(PRECISION).div(stSolSupply);
     }
-    throw new Error("UnsupportedBasePool");
+    throw new Error('UnsupportedBasePool');
   }
 
-  private updateDepegInfoIfExpired() {
-    if (!this.depeg.depegType["none"]) {
-      const expired =
-        this.onChainTime >
-        this.depeg.baseCacheUpdated.add(BASE_CACHE_EXPIRE).toNumber();
+  private async updateDepegInfoIfExpired() {
+    const onChainTime = await getOnchainTime(this.connection);
```

Comment:
`onChainTime` will be used only when `expire`, we can move it into `if (expired)` block to save 1 RPC call